blobfuse/OAuthTokenCredentialManager.cpp

```diff
@@ -0,0 +1,216 @@
+{
+    try {
+        current_oauth_token = refreshTokenCallback(httpClient);
+        return current_oauth_token;
+    } catch(std::exception& ex) {
+        valid_authentication = false;
+        throw ex;
+    }
+}
+
+/// <summary>
+/// Returns current oauth_token, implicitly refreshing if the current token is invalid.
+/// Note that this can throw an error if the refresh fails, so be prepared to catch.
+/// </summary>
+OAuthToken OAuthTokenCredentialManager::get_token()
+{
+    if (is_token_expired()) {
+        if (token_mutex.try_lock()) {
+            try {
+                refresh_token();
+            } catch (std::exception &ex) {
+                syslog(LOG_ERR, "Unable to retrieve OAuth token: %s", ex.what());
+                valid_authentication = false;
+                throw std::runtime_error(std::string("Failed to refresh OAuth token: ") + std::string(ex.what()));
+            }
+        } else {
+            time_t current_time;
+            time(&current_time);
+
+            // There's a five minute segment where the token hasn't actually expired yet, but we're already trying to refresh it.
+            // We can just use the currently active token instead, rather than waiting for the refresh.
+            // This'll save some downtime on systems under constant use.
+            if (current_time < current_oauth_token.expires_on) {
+                return current_oauth_token;
+            } else {
+                // If it's not still live, let's just wait for the refresh to finish.
+                // This is a sub-optimal method to wait for this event as it can end up blocking other routines after the token has finished refreshing.
+                // If we were working in Go, I (Adele) would suggest using a sync.WaitGroup for this functionality.
+                token_mutex.lock();
+            }
+        }
+        token_mutex.unlock();
+    }
+
+    return current_oauth_token;
+}
+
+/// <summary>
+/// TODO: check the expiry time against the current utc time
+/// <summary>
+bool OAuthTokenCredentialManager::is_token_expired()
+{
+    time_t current_time;
+
+    time ( &current_time );
+
+    // check if about to expire via the buffered expiry time
+    return current_time + (60 * 5) >= current_oauth_token.expires_on;
+}
+
+// ===== CALLBACK SETUP ZONE =====
+
+/// <summary>
+/// SetUpMSICallback sets up a refresh callback for MSI auth. This should be used to create a OAuthTokenManager instance.
+/// </summary>
+std::function<OAuthToken(std::shared_ptr<CurlEasyClient>)> SetUpMSICallback(std::string client_id_p, std::string object_id_p, std::string resource_id_p)
+{
+    // Create the URI token request
+    std::shared_ptr<microsoft_azure::storage::storage_url> uri_token_request_url = std::make_shared<microsoft_azure::storage::storage_url>();
+    uri_token_request_url->set_domain(constants::msi_request_uri);
+    uri_token_request_url->append_path(constants::msi_request_path);

```

Comment:
Can you combine msi_request_uri and msi_request_path and instead use a single variable? The default constant should be used only if the environment variable MSI_ENDPOINT is set. Consider using the same approach as [here](https://github.com/Azure/azure-sdk-for-net/blob/b0ff97d220ec1e83d43767d693b17a3b6245314e/sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs#L87) to construct this endpoint url.